src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs

```diff
@@ -123,9 +126,61 @@ protected override void RemoveItem(int index)
 
             OnCountPropertyChanged();
             OnIndexerPropertyChanged();
-            OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItem, index);
+
+            if (!_skipCollectionChanged)
+            {
+                OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItem, index);
+            }
         }
 
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a count of items is removed from the list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void RemoveItemsRange(int index, int count)
+        {
+            CheckReentrancy();
+
+            int length = this.Count;
+            if (index < this.Count - count)
+            {
+                length = index + count;
+            }
+
+            T[] removedItems = new T[length];
+            for (int i = index; i < length; i++)
+            {
+                removedItems[i] = this[index];
+            }
+
+            _skipCollectionChanged = true;
+            try
+            {
+                base.RemoveItemsRange(index, count);
+            }
+            finally
+            {
+                _skipCollectionChanged = false;
+            }
+
+            OnCountPropertyChanged();
+            OnIndexerPropertyChanged();
+            OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItems, index);
+        }
+
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a collection of items is added to list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void ReplaceItemsRange(int index, int count, IEnumerable<T> collection)
+        {
+            CheckReentrancy();
+            base.ReplaceItemsRange(index, count, collection);
```

Comment:
Isn't this going to result in more events raised than necessary?

The default implementation of `base.ReplaceItemsRange` calls `RemoveItemsRange` and `InsertItemsRange` which each call `OnCountPropertyChanged`, `OnIndexerPropertyChanged`, and `OnCollectionChanged `, then this is calling `OnCountPropertyChanged` and `OnIndexerPropertyChanged` itself.

To summarize the calls:
 - `OnCountPropertyChanged()`
 - `OnIndexerPropertyChanged()`
 - `OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItems, index)`
 - `OnCountPropertyChanged()`
 - `OnIndexerPropertyChanged()`
 - `OnCollectionChanged(NotifyCollectionChangedAction.Add, collection, index)`
 - `OnCountPropertyChanged()`
 - `OnIndexerPropertyChanged()`

Shouldn't it be:
 - `OnCountPropertyChanged()`
 - `OnIndexerPropertyChanged()`
 - `OnCollectionChanged(NotifyCollectionChangedAction.Replace, newItems, oldItems, index)`

Note the last bullet is raising a single `NotifyCollectionChangedAction.Replace` event instead of separate `NotifyCollectionChangedAction.Remove` and `NotifyCollectionChangedAction.Add` events, which I _think_ is what we want (but I could be wrong).

Note that the bullet written above requires a new `OnCollectionChanged` helper overload to call the following `NotifyCollectionChangedEventArgs` constructor:

https://github.com/dotnet/corefx/blob/b66846b62fd39973eb54204087fe311941c87235/src/System.ObjectModel/src/System/Collections/Specialized/NotifyCollectionChangedEventArgs.cs#L231-L254

Also note that it isn't as simple as passing the passed-in `collection` to that `NotifyCollectionChangedEventArgs` constructor for the `newItems` parameter as it's typed as `IList` and `collection` is `IEnumerable<T>`, so `collection` would have to be cast to `IList` (if it is an `IList`), otherwise it'd need to be copied to an array.

Reply:
@justinvp this is a great write up of the problem and it currently does send extra notifications.

I am going to do some tests to see how we can best implement this. My biggest concern is copying an array may be very expensive especially if T is a large class and the underlying array is close to `int.MaxValue`

I'll report back with my findings and what I think would be best to solve this problem.

Reply:
> I am going to do some tests to see how we can best implement this. My biggest concern is copying an array may be very expensive especially if T is a large class and the underlying array is close to `int.MaxValue`

It's probably simplest to do something like:

```c#
IList newItems = collection is IList list ?
    list :
    new List<T>(collection);
```

Reply:
@justinvp I have updated the `ReplaceItemsRange` to only fire the 3 events and pass the correct event arguments.